GVFS/GVFS.Common/GVFSEnlistmentStatistics.cs

```diff
@@ -0,0 +1,202 @@
+        }
+
+        public double GetPlaceholderPercentage()
+        {
+            if (this.GitTrackedFilesCount == 0)
+            {
+                return 0;
+            }
+
+            return (double)this.PlaceholderCount / this.GitTrackedFilesCount;
+        }
+
+        public double GetModifiedPathsPercentage()
+        {
+            if (this.GitTrackedFilesCount == 0)
+            {
+                return 0;
+            }
+
+            return (double)this.ModifiedPathsCount / this.GitTrackedFilesCount;
+        }
+
+        /// <summary>
+        /// Take a string representing a file path on system and pull out the upper-most directory from it as a string, or GVFSConstants.GitPathSeparator if it is in the root
+        /// </summary>
+        /// <param name="path">The path to a file to parse for the top level directory containing it</param>
+        /// <returns>A string containing the top level directory from the provided path, or GVFSConstants.GitPathSeparator if the path is for an item in the root</returns>
+        private string ParseTopDirectory(string path)
+        {
+            int whackLocation = path.IndexOf(GVFSConstants.GitPathSeparator);
+            if (whackLocation == -1)
+            {
+                return GVFSConstants.GitPathSeparator.ToString();
+            }
+
+            return path.Substring(0, whackLocation);
+        }
+
+        /// <summary>
+        /// Categorizes a list of paths given as strings by mapping them to the top level directory in their path
+        /// Modifies the directoryTracking dictionary to have an accurate count of the files underneath a top level directory
+        /// </summary>
+        /// <remarks>
+        /// The distinction between files and directories is important --
+        /// If the path to a file doesn't contain a GVFSConstants.GitPathSeparator, then that means it falls within the root
+        /// However if a directory's path doesn't contain a GVFSConstants.GitPathSeparator, it counts towards itself for hydration
+        /// </remarks>
+        /// <param name="paths">An enumerable containing paths as strings</param>
+        /// <param name="directoryTracking">A dictionary used to track the number of files per top level directory</param>
+        private int CategorizePaths(IEnumerable<string> paths, Dictionary<string, int> directoryTracking, string parentDirectory, bool isFile)
+        {
+            int count = 0;
+            foreach (string path in paths)
+            {
+                // Ensure all paths have the same slashes
+                string formattedPath = path.Replace('\\', GVFSConstants.GitPathSeparator);
+                if (formattedPath.StartsWith(GVFSConstants.GitPathSeparator.ToString()))
+                {
+                    formattedPath = formattedPath.Substring(1);
+                }
+
+                if (formattedPath.StartsWith(parentDirectory))

```

Comment:
Is there a trailing delimiter to the `parentDirectory`?  If not then this is a bug because passing in `foo`  will catch everything from files named `foo..` or directories named `foo...` If we only want the `foo` directory we will need the `/` as a part of the parentDirectory so that `StartsWith` will not get those other items.

Reply:
Great catch! I'll add the check into `CalculateStatistics()` where it's initially passed in. This way regardless of whether or not they have one we'll ensure it actually only sees children of the directory

Reply:
I'm adding a unit test for this case as well!